### PR TITLE
MultiPolygon: isOGCValid and isOSMValid

### PR DESCRIPTION
**Pull request by matthieun**
### Description:

The OGC definition of validity for multipolygons (backed by JTS) slightly differs from the one in OpenStreetMap. When all OGC valid multipolygons can have inners that touch each other at most at one point, OSM allows two inners to hug each other along a set of linear segments, to allow for separate tagging within some multipolygon inners.

Also:
- to facilitate that work, now `GeometryOperation` supports computing the intersection of a list of `Polygon`s.

### Potential Impact:

These are validity test methods, so there is no impact so far.

### Unit Test Approach:

Added a bunch of new tests with mock multipolygons.

### Test Results:

Pass

------

In doubt: [Contributing Guidelines](CONTRIBUTING.md)

### CODE REVIEW

File: src/main/java/org/openstreetmap/atlas/geography/MultiPolygon.java [hunk 1]
```diff
@@ -386,6 +395,47 @@ public boolean isEmpty()
...
+        {
+            // In this case, the geometry is not OGC valid, here we capture the
+            // TopologyValidationError to know what to do next.
+            if (TopologyValidationError.SELF_INTERSECTION == topologyValidationError.getErrorType())
+            {
+                final Location errorLocation = new JtsLocationConverter()
+                        .backwardConvert(topologyValidationError.getCoordinate());
+                final Rectangle errorExpandedBoundingBox = errorLocation
+                        .boxAround(Distance.ONE_METER);
+                return isOSMValidSelfIntersection(errorExpandedBoundingBox);
```
Comment (adahn6) [inline src/main/java/org/openstreetmap/atlas/geography/MultiPolygon.java:429]: oh this is cool!

File: src/test/java/org/openstreetmap/atlas/geography/MultiPolygonTest.java [hunk 2]
```diff
@@ -204,4 +221,101 @@ public void testSurface()
...
+    }
+
+    @Test
+    public void testValidity9()
+    {
+        // Inner outside outer, hugging it along a segment
+        final MultiPolygon multiPolygon = getFrom("MultiPolygonTestGeom9.josm.osm");
+        Assert.assertFalse(multiPolygon.isOGCValid());
+        Assert.assertFalse(multiPolygon.isOSMValid());
+    }
```
Comment (adahn6) [inline src/test/java/org/openstreetmap/atlas/geography/MultiPolygonTest.java:320]: nice tests!!

File: src/main/java/org/openstreetmap/atlas/geography/MultiPolygon.java [hunk 3]
```diff
@@ -386,6 +395,47 @@ public boolean isEmpty()
...
+    {
+        final org.locationtech.jts.geom.MultiPolygon jtsMultiPolygon = new JtsMultiPolygonToMultiPolygonConverter()
+                .backwardConvert(this);
+        final TopologyValidationError topologyValidationError = new IsValidOp(jtsMultiPolygon)
+                .getValidationError();
+        if (topologyValidationError != null)
+        {
+            // In this case, the geometry is not OGC valid, here we capture the
+            // TopologyValidationError to know what to do next.
+            if (TopologyValidationError.SELF_INTERSECTION == topologyValidationError.getErrorType())
```
[the remark this answers is not in the document]
Comment (matthieun) [reply src/main/java/org/openstreetmap/atlas/geography/MultiPolygon.java:423]: Both are an int primitive...

[the remark this answers is not in the document]
Comment (adahn6) [reply src/main/java/org/openstreetmap/atlas/geography/MultiPolygon.java:423]: OH that's what I get for not looking at what the values were, lol

**Pull request merged**